### PR TITLE
Adds support for changing passwords in the Account controller

### LINKED ISSUE #423: Change Password
Create the change password page.

### PR DESCRIPTION
**Pull request by lukaskabrt**
Adds support for changing passwords in the Account controller. The same rules for password complexity as in the registration apply.

Fixes #423

### CODE REVIEW

File: src/OrchardCore.Modules/Orchard.Users/Views/Account/ChangePassword.cshtml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿@model Orchard.Users.ViewModels.ChangePasswordViewModel
+
+<form asp-controller="Account" asp-action="ChangePassword" method="post" class="form-horizontal">
+    <h4>@T["Change password"]</h4>
+    <hr />
```
Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Views/Account/ChangePassword.cshtml]: Not sure we use `<hr/>` in other views. If not then remove it.

File: src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs [hunk 2]
```diff
@@ -124,6 +124,28 @@ public IActionResult Register(string returnUrl = null)
...
+        public IActionResult ChangePassword()
+        {
+            return View();
+        }
+
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<IActionResult> ChangePassword(ChangePasswordViewModel model)
+        {
+            if (ModelState.IsValid)
```
Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: Shouldn't it check for the user to be authenticated first? Maybe with an [Authorize] attribute on the action.

Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: Forget about it, it's already on the controller

File: src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs [hunk 3]
```diff
@@ -124,6 +124,28 @@ public IActionResult Register(string returnUrl = null)
...
+            {
+                var user = await _userService.GetAuthenticatedUserAsync(User);
+                if (await _userService.ChangePasswordAsync(user, model.CurrentPassword, model.Password, (key, message) => ModelState.AddModelError(key, message)))
+                {
+                    return View("ChangePasswordSuccess");
```
Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: It should return a RedirectLocal because it's a POST

Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: [HttpGet]
        public IActionResult ChangePasswordConfirmation()
        {
            return View();
        }

File: src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs (hunk 2 above)
Comment (lukaskabrt) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: The whole Account controller is marked with the [Authorize] attribute, so it shouldn't be neccessary

Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: We agree ;)

File: src/OrchardCore.Modules/Orchard.Users/Views/Account/ChangePassword.cshtml (hunk 1 above)
Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Views/Account/ChangePassword.cshtml]: Good then

Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Views/Account/ChangePassword.cshtml]: Thanks for following the existing templates

File: src/OrchardCore.Modules/Orchard.Users/Services/UserService.cs [hunk 4]
```diff
@@ -49,55 +53,83 @@ public UserService(UserManager<User> userManager, IOptions<IdentityOptions> iden
...
+        {
+            if(principal == null)
+            {
+                return null;
+            }
+
+            return await _userManager.GetUserAsync(principal);
+        }
+
+        private void ProcessIdentityErrors(IEnumerable<IdentityError> errors, User user, Action<string, string> reportError)
```
Comment (lukaskabrt) [inline src/OrchardCore.Modules/Orchard.Users/Services/UserService.cs]: I am not a native speaker, but the name `ValidateUser` gives me impression that the method actually preforms some validation of the `user` parameter. In fact the validation is done by the `UserManager` class. What about `ProcessValidationErrors`?

Of course, if I am wrong, I will change the name to `ValidateUser` ...

File: src/OrchardCore.Modules/Orchard.Users/Services/UserService.cs [hunk 5]
```diff
@@ -49,55 +53,83 @@ public UserService(UserManager<User> userManager, IOptions<IdentityOptions> iden
...
         }
+
+        public async Task<User> GetAuthenticatedUserAsync(ClaimsPrincipal principal)
+        {
+            if(principal == null)
+            {
+                return null;
+            }
+
+            return await _userManager.GetUserAsync(principal);
```
Comment (lukaskabrt) [inline src/OrchardCore.Modules/Orchard.Users/Services/UserService.cs]: fixed

File: src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs (hunk 3 above)
Comment (lukaskabrt) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: fixed, thanks for catching this

Comment (sebastienros) [inline src/OrchardCore.Modules/Orchard.Users/Controllers/AccountController.cs]: if you return to a local action, just use RedirectToAction

**Pull request merged**